src/java.base/share/classes/java/lang/Float16.java

```diff
@@ -56,129 +73,941 @@
-     * @since 20
+     * Returns the smaller of two {@code Float16} values.
+     *
+     * @apiNote
+     * This method corresponds to the minimum operation defined in
+     * IEEE 754.
+     *
+     * @param a the first operand
+     * @param b the second operand
+     * @return the smaller of {@code a} and {@code b}
+     * @see java.util.function.BinaryOperator
+     * @see Math#min(float, float)
+     * @see Math#min(double, double)
+     */
+    public static Float16 min(Float16 a, Float16 b) {
+        return shortBitsToFloat16(floatToFloat16(Math.min(a.floatValue(),
+                                                          b.floatValue()) ));
+    }
+
+    // Skipping for now
+    // public Optional<Float16> describeConstable()
+    // public Float16 resolveConstantDesc(MethodHandles.Lookup lookup)
+
+    /*
+     * Note: for the basic arithmetic operations {+, -, *, /} and
+     * square root, among binary interchange formats (binary16,
+     * binary32 a.k.a. float, binary64 a.k.a double, etc.) the "2p + 2"
+     * property holds. That is, if one format has p bits of precision,
+     * if the next larger format has at least 2p + 2 bits of
+     * precision, arithmetic on the smaller format can be implemented by:
+     *
+     * 1) converting each argument to the wider format
+     * 2) performing the operation in the wider format
+     * 3) converting the result from 2) to the narrower format
+     *
+     * For example, this property hold between the formats used for the
+     * float and double types. Therefore, the following is a valid
+     * implementation of a float addition:
+     *
+     * float add(float addend, float augend) {
+     *     return (float)((double)addend + (double)augend);
+     * }
+     *
+     * The same property holds between the float16 format and
+     * float. Therefore, the software implementations of Float16 {+,
+     * -, *, /} and square root below use the technique of widening
+     * the Float16 arguments to float, performing the operation in
+     * float arithmetic, and then rounding the float result to
+     * Float16.
+     */
+
+    /**
+     * Adds two {@code Float16} values together as per the {@code +}
+     * operator semantics using the round to nearest rounding policy.
+     *
+     * The handling of signed zeros, NaNs, infinities, and other
+     * special cases by this method is the same as for the handling of
+     * those cases by the built-in {@code +} operator for
+     * floating-point addition (JLS {@jls 15.18.2}).
+     *
+     * @apiNote This method corresponds to the addition operation
+     * defined in IEEE 754.
+     *
+     * @param addend the first operand
+     * @param augend the second operand
+     * @return the sum of the operands
+     *
+     * @jls 15.4 Floating-point Expressions
      */
-    public short float16ToRawShortBits() { return value; }
+    // @IntrinsicCandidate
+    public static Float16 add(Float16 addend, Float16 augend) {
```

Comment:
Hi, thanks for implementing these operations for Float16. May I ask why do we have both `Float16 sum()` and `Float16 add()`? Also, both seem to be marked as `@IntrinsicsCandidate`.

Reply:
> Hi, thanks for implementing these operations for Float16. May I ask why do we have both `Float16 sum()` and `Float16 add()`? Also, both seem to be marked as `@IntrinsicsCandidate`.

Hello @Bhavana-Kilambi ,
The Float16 class already had a sum method before I started my work, so I left it in place. I assume "sum" was added in analogy to the the sum method added to Float and Double in the context of the stream work done in JDK 8.

For the purpose of including support for methods to perform the addition, subtraction, multiplication, and division operation, I think "add" is a better name than "sum".

I don't have any appreciable experience changing HotSpot, so I was going to leave that aspect of this work to others. The build fails if the `@IntrinsicCandidate` annotation is applied to methods without the possibility of a matching candidate in HotSpot. Therefore, I left a commented out form of the annotation on methods I assumed would be eligible for such support, but those were just my guesses and should be updated for what the actual intrinsic support could be.

Reply:
Ok, got it. Thank you for the detailed response.